File: testdomains/src/main/java/overflowdb/testdomains/gratefuldead/ArtistDb.java
package overflowdb.testdomains.gratefuldead;

import overflowdb.NodeLayoutInformation;
import overflowdb.NodeRef;
import overflowdb.NodeDb;

import java.util.Arrays;
import java.util.HashMap;
import java.util.HashSet;
import java.util.Map;

public class ArtistDb extends NodeDb {
  protected ArtistDb(NodeRef ref) {
    super(ref);
  }

  private String _name;

  public String name() {
    return _name;
  }

  @Override
  public NodeLayoutInformation layoutInformation() {
    return layoutInformation;
  }

  @Override
  public Object property(String key) {
    if (Artist.NAME.equals(key)) {
      return _name;
    } else {
      return null;
    }
  }

  @Override
  protected void updateSpecificProperty(String key, Object value) {
    if (Artist.NAME.equals(key)) {
      this._name = (String) value;
    } else {
      throw new RuntimeException("property with key=" + key + " not (yet) supported by " + this.getClass().getName());
    }
  }

  @Override
  protected void removeSpecificProperty(String key) {
    if (Artist.NAME.equals(key)) {
      this._name = null;
    } else {
      throw new RuntimeException("property with key=" + key + " not (yet) supported by " + this.getClass().getName());
    }
  }

  public static NodeLayoutInformation layoutInformation = new NodeLayoutInformation(
      Artist.label,
      new HashSet<>(Arrays.asList(Artist.NAME)),
      Arrays.asList(),
      Arrays.asList(SungBy.layoutInformation, WrittenBy.layoutInformation));
}

File: traversal-tests/src/test/scala/overflowdb/traversal/testdomains/gratefuldead/ArtistDb.scala
package overflowdb.traversal.testdomains.gratefuldead

import overflowdb.traversal._
import overflowdb.{NodeRef, NodeDb}

class ArtistDb(ref: NodeRef[ArtistDb]) extends NodeDb(ref) {
  /* name property */
  def name: String = _name
  private var _name: String = null

  /* Artist <-- sungBy --- Song */
  def sangSongs: Traversal[Song] = in(SungBy.Label).toScalaAs[Song]

  /* Artist <-- writtenBy --- Song */
  def wroteSongs: Traversal[Song] = in(WrittenBy.Label).toScalaAs[Song]

  override def property(key: String) =
    key match {
      case Artist.PropertyNames.Name => _name
      case _                         => null
    }

  override protected def updateSpecificProperty(key: String, value: Object) =
    key match {
      case Artist.PropertyNames.Name =>
        _name = value.asInstanceOf[String]
        property(Artist.PropertyNames.Name)
      case _ =>
        throw new RuntimeException("property with key=" + key + " not (yet) supported by " + this.getClass().getName());
    }

  override protected def removeSpecificProperty(key: String) =
    key match {
      case Artist.PropertyNames.Name =>
        _name = null
      case _ =>
        throw new RuntimeException("property with key=" + key + " not (yet) supported by " + this.getClass().getName());
    }

  override def layoutInformation = Artist.layoutInformation
}
